### PR TITLE
fix: add the OIDC handler for device flow

### PR DESCRIPTION
**Pull request by wood-push-melon**
This pull request tries to:

- Add the OIDC handler for the device authorization grant
- Add the handler tests

### CODE REVIEW

File: handler/openid/flow_device_auth.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+		return nil
+	}
+
+	if len(resp.GetDeviceCode()) == 0 {
+		return errorsx.WithStack(fosite.ErrMisconfiguration.WithDebug("The device code has not been issued yet, indicating a broken code configuration."))
+	}
+
+	if err := c.OpenIDConnectRequestValidator.ValidatePrompt(ctx, dar); err != nil {
+		return err
+	}
```
Comment (nsklikas) [inline handler/openid/flow_device_auth.go]: Not sure if we want this, a `prompt` param is not defined in the device flow spec (it makes sense as it is an oauth2 rfc and `prompt` is an oidc param).

I took a look at how this is handled by google, microsoft and amazon and they do not talk about a `prompt` param being supported in their docs. I would suggest to remove this

File: handler/openid/validator.go
Comment (nsklikas) [inline handler/openid/validator.go:1]: We can undo these changes, if we remove the prompt validation from the device auth request handling

File: handler/openid/flow_device_auth.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+func (c *OpenIDConnectDeviceHandler) HandleDeviceEndpointRequest(ctx context.Context, dar fosite.DeviceRequester, resp fosite.DeviceResponder) error {
+	if !(dar.GetGrantedScopes().Has("openid")) {
+		return nil
+	}
+
+	if !dar.GetClient().GetGrantTypes().Has(string(fosite.GrantTypeDeviceCode)) {
+		return nil
+	}
```
Comment (nsklikas) [inline handler/openid/flow_device_auth.go:34]: Not sure about this but I don't think this is needed, the oauth2 handler will have already checked this.

Feel free to ignore this comment as it looks like fosite is not very consistent on what checks each handler needs to implement

**Review by nsklikas — COMMENTED** [review]
lgtm overall, I have a couple of minor comments. 

I think that we should remove the prompt parsing, the other comment is a nit-pick and you can ignore it if you disagree

File: handler/openid/flow_device_auth.go (hunk 1 above)
Comment (wood-push-melon) [reply handler/openid/flow_device_auth.go]: Yeah, you are right. The `prompt` parameter is less relevant in device flow because of the user interaction separation. The `prompt` does not do any help in the initial authorization request.

File: handler/openid/flow_device_auth.go (hunk 2 above)
Comment (wood-push-melon) [reply handler/openid/flow_device_auth.go:34]: Yeah, I am also confused by lots of condition checks in the upstream. It seems that they've done several duplicated checks in handling requests with handlers in a sequential order.

**Review by nsklikas — APPROVED** [review]

File: handler/openid/flow_device_auth.go [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+	if len(resp.GetDeviceCode()) == 0 {
```
Comment (shipperizer) [inline handler/openid/flow_device_auth.go:36]:
```suggestion
	if resp.GetDeviceCode() == "" {
```

small q: why are we checking the lenght instead of the emptiness?

File: handler/openid/flow_device_token.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+	claims := session.IDTokenClaims()
+	if claims.Subject == "" {
+		return errorsx.WithStack(fosite.ErrServerError.WithDebug("Failed to generate id token because subject is an empty string."))
+	}
+
+	claims.AccessTokenHash = c.GetAccessTokenHash(ctx, requester, responder)
+
+	idTokenLifespan := fosite.GetEffectiveLifespan(requester.GetClient(), fosite.GrantTypeDeviceCode, fosite.IDToken, c.Config.GetIDTokenLifespan(ctx))
+	return c.IssueExplicitIDToken(ctx, idTokenLifespan, ar, responder)
```
Comment (shipperizer) [inline handler/openid/flow_device_token.go:55]: image

took me a rocket scientist class to understand what's going on here

looks good anyway

**Review by shipperizer — APPROVED** [review]
squash the commits into 1 and it's good to go imho

File: handler/openid/flow_device_auth.go (hunk 3 above)
Comment (wood-push-melon) [reply handler/openid/flow_device_auth.go:36]: Ah, copied from the upstream. Will change it in another bug-fix PR. Thanks.
